Optimise GHAppy to use Text instead of String

kpadmasola opened the pull request:
+ Resolves https://github.com/mlabs-haskell/GHAppy/issues/14

kpadmasola committed:
Optimise GHAppy to use Text instead of String

+ Resolves https://github.com/mlabs-haskell/GHAppy/issues/14

kpadmasola commented:
@cstml I am a beginner in Haskell, and came across this issue in Haskell Weekly. The changes compile, tests are passing and pre-commit checks are green. However I am not sure if I have addressed the changes in the most efficient manner as there are still some places where I had to do pack/unpack. Please let me know if the changes need to be revised and I'll try to address the comments. Thanks!

cstml commented:
Hi @kpadmasola,

I recommend you:
1. add `OverloadedStrings` as an extension to `GHAppy.cabal` - and then you might get rid of some of the packing that you need to do. I'll add some more comments to the PR.
2. Be sure to capture all of the places where `String` is used - [here](https://github.com/placidex/GHAppy/blob/49f569c2b02c89d44245dbd96990a7e9b95daa43/src/GHAppy.hs#L374) for example.

I'll add a few more comments - but thank you for your contribution - much appreciated.

cstml commented on src/GHAppy.hs:
```diff
@@ -301,7 +306,7 @@ compose = reinterpret go
         let f = fmap ((\n -> emptyLeaf {issueN = n, level = lvl}) . Just . fst) . M.toList . M.filter (getPredicate p) . unIssues
          in (get >>= tell . f)
-      AddHeader lvl str -> tell [emptyLeaf {preamble = [replicate (fromEnum lvl) '#' <> " " <> str]}]
+      AddHeader lvl str -> tell [emptyLeaf {preamble = [T.pack $ replicate (fromEnum lvl) '#' <> " " <> T.unpack str]}]
```
```suggestion
      AddHeader lvl str -> tell [emptyLeaf {preamble = [T.pack (replicate (fromEnum lvl) '#') <> " " <> str]}]
```
If you move the bracket you avoid packing and unpacking - also it helps to have that `OverloadedStrings` on

cstml commented on app/Example.hs:
```diff
@@ -19,8 +21,8 @@ main = do
       let logo = "MLabs-logo-cropped" <.> "jpg"
       let logoCropped = "MLabs-logo" <.> "jpg"
       let linkedFiles = "https://raw.githubusercontent.com/mlabs-haskell/audit-report-template/master/linked-files/images/"
```
Packing these in the let binding would be a bit clearer. 
```suggestion
      let logo = T.pack $ "MLabs-logo-cropped" <.> "jpg"
      let logoCropped = T.pack $ "MLabs-logo" <.> "jpg"
      let linkedFiles = "https://raw.githubusercontent.com/mlabs-haskell/audit-report-template/master/linked-files/images/"
```

cstml reviewed (changes requested):
Hi @kpadmasola,

I recommend you:
1. add `OverloadedStrings` as an extension to `GHAppy.cabal` - and then you might get rid of some of the packing that you need to do. I'll add some more comments to the PR.
2. Be sure to capture all of the places where `String` is used - [here](https://github.com/placidex/GHAppy/blob/49f569c2b02c89d44245dbd96990a7e9b95daa43/src/GHAppy.hs#L374) for example.

Thank you for your contribution - much appreciated.

kpadmasola commented:
> Hi @kpadmasola,
> 
> I recommend you:
> 
>     1. add `OverloadedStrings` as an extension to `GHAppy.cabal` - and then you might get rid of some of the packing that you need to do. I'll add some more comments to the PR.
> 

Hi @cstml , `OverloadedStrings` is already present in `GHAppy.cabal` at https://github.com/placidex/GHAppy/blob/49f569c2b02c89d44245dbd96990a7e9b95daa43/GHAppy.cabal#L45. Is there some other step that is missing?

>     2. Be sure to capture all of the places where `String` is used - [here](https://github.com/placidex/GHAppy/blob/49f569c2b02c89d44245dbd96990a7e9b95daa43/src/GHAppy.hs#L374) for example.
> 

Thanks, initially I tried that, but I got overwhelmed by the errors. I'll try again.
By the way in case I run into any further issues are you reachable on discord?

> 
> Thank you for your contribution - much appreciated.

😄

kpadmasola committed:
Update src/GHAppy.hs

Co-authored-by: Vlad <cstmlcodes@gmail.com>

kpadmasola committed:
Update app/Example.hs

Co-authored-by: Vlad <cstmlcodes@gmail.com>

cstml commented:
> Hi @cstml , OverloadedStrings is already present in GHAppy.cabal

Great - my oversight - then you should leverage that and allow the packing to happen by itself in some cases.

> Thanks, initially I tried that, but I got overwhelmed by the errors. I'll try again.

No worries - start in one place and fix errors, then start in the next place, etc. 

> By the way in case I run into any further issues are you reachable on discord?

You can reach out to me on matrix at `@cstm:matrix.org`. I might reply faster here tbh.

kpadmasola committed:
Use Text instead of String in more places

kpadmasola committed:
Cleanup YamlInterface

kpadmasola commented:
@cstml I have addressed the review comments, please take a look when you get a chance. Thanks!

kpadmasola committed:
fix for CI failure

+ When running `nix build` it fails with `Error: missing or private
  dependencies: pandoc >=2.19.0 && <3`. Removing the lower bound
  makes `nix build` pass. Not sure of the root cause though.

kpadmasola commented:
@cstml I pushed a fix for the CI build failure, please take a look. The CI build is not getting triggered for this PR automatically, perhaps because I am a first time contributor. Locally
however, I got the `nix build` to pass, so I am hopeful it will pass in the CI as well.
